### ISSUE TITLE
Question about lexicon_features and vectorize

### ISSUE THREAD

**Issue by qsong4** [iit-cs579/main#248]
The function lexicon_features return the result like [('neg_words', 1), ('pos_words', 2)], so at the function vectorize,  how to sort the columns alphabetically? For example:

    [('neg_words', 0), ('pos_words', 1), ('token=great', 2), ('token=horrible', 3), ('token=isn', 4), ('token=movie', 5), ('token=t', 6), ('token=this', 7)]

Is this the correct order?

**Comment by vjonnala** [comment]
In function vectorize(), you need to sort based on feature count items. For example:
[('token=0', 0), ('token=1', 1), ('token=1/2', 2), ('token=10', 3), ('token=10/10', 4), ('token=100', 5), ('token=11', 6), ('token=12', 7), ('token=15', 8), ('token=16', 9)]
In function featurize(), you need to compute all the feats and sort its items.

Hope this helps.

**Comment by qsong4** [comment]
@vjonnala 
      - vocab: a dict from feature name to column index. NOTE
      that the columns are sorted alphabetically (so, the feature
      "token=great" is column 0 and "token=horrible" is column 1
      because "great" < "horrible" alphabetically)
So, for the function vectorize(), if we want to sort the column we need to add 'token=' in front of 'neg_words' and 'pos_words'. Correct me if I'm wrong.

**Comment by vjonnala** [comment]
Your vocab contains 'token='. Please see the notes of vectorize() method.

**Comment by viyatgandhi** [comment]
@vjonnala @ClaytonTurner @aronwc 
I am also confused with lexicon_feature with vectorization method:

Scenario 1:
Input
docs = ["Isn't this movie great?", "Horrible, horrible movie"]
    tokens_list = [tokenize(d) for d in docs]
    feature_fns = [lexicon_features]
    vectorize(tokens_list, feature_fns, min_freq=1)

Output of matrix will be 
[[0 1]
 [2 0]]
vocab will be
[('neg_words', 0), ('pos_words', 1)]

Scenario 2:
Consider min_freq=2 - Does it mean we need to check pos_words and neg_words occurrence 
in different documents for lexicon_feature ?

Input
docs = ["Isn't this movie great?", "Horrible, horrible movie"]
    tokens_list = [tokenize(d) for d in docs]
    feature_fns = [lexicon_features]
    vectorize(tokens_list, feature_fns, min_freq=2)

Output of matrix will be 
[[0 0]
 [0 0]]

vocab will be
[('neg_words', 0), ('pos_words', 1)]

As pos+ and neg- words does not occur in two docs we should not consider it ?

Scenario 3:
As mentioned in original issue
Input 
docs = ["Isn't this movie great?", "Horrible, horrible movie"]
    tokens_list = [tokenize(d) for d in docs]
    feature_fns = [token_features,lexicon_features]
    vectorize(tokens_list, feature_fns, min_freq=1)

Output of matrix will be 
[[0 1 1 0 1 1 1 1]
 [2 0 0 2 0 1 0 0]]

vocab will be
[('neg_words', 0), ('pos_words', 1), ('token=great', 2), ('token=horrible', 3), ('token=isn', 4), ('token=movie', 5), ('token=t', 6), ('token=this', 7)]

TR;DR 
How does lexicon_features works with min_freq
and How does lexicon_features works with in combination with other features for vectorization?

Hope question is clear to all and my doubts are legitimate.

**Comment by viyatgandhi** [comment]
@vjonnala @ClaytonTurner @aronwc 
Can you please clarify above doubt ? It will expedite my assignment.

**Comment by davidghiurco** [comment]
as per vectorize() description:

      min_freq......Remove features that do not appear in
                    at least min_freq different documents.

in short, min_freqs checks to see the number of documents that feature x appears in. it is NOT descriptive of the total frequency of a certain feature.

lexicon_features is just like any other features. So yes, if post_words or neg_words does not appear at least min_freq times in different documents you do not consider it. Your scenario 2 is correct. 

Long story short: Again, lexicon_features is just a feature like any other. it doesn't need any special consideration over the other features. All features need to be checked for min_freq the same way.
Hint: have a counter that counts the number of documents each feature appears in. THIS IS NOT THE SAME AS THE FREQUENCY OF THAT FEATURE IN ALL DOCUMENTS****

**Comment by davidghiurco** [comment]
also @qsong4  you are correct. The alphabetical ordering is according to neg_words, pos_words, and token=

meaning that all tokens are alphabetized starting with "token=" and since neg_words and pos_words start with n, and p, respectively, those both come before the t in "token="

**Comment by viyatgandhi** [comment]
@davidghiurco thanks for clarification. @qsong4 you can close this if your problem is resolved.
